### PR TITLE
Add option to disable attachment pages

### PR DESCRIPTION
**Pull request by aristath**
This PR does the following:
* Adds an `attachment-pages` theme-support
* Redirects (`301` redirect by default) attachment pages to the attachment-URLs (direct link to the file itself)

Trac ticket: https://core.trac.wordpress.org/ticket/57913

props @carolinan @SergeyBiryukov @aristath

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/includes/upgrade.php [hunk 1]
```diff
@@ -2290,6 +2295,23 @@ function upgrade_600() {
...
+ * @global int $wp_current_db_version The old (current) database version.
+ */
+function upgrade_630() {
+	global $wp_current_db_version;
+
+	// @TODO: Change the number to something that makes sense. This is random for the sake of this POC.
+	if ( $wp_current_db_version < 54248 ) {
+		update_option( 'wp_media_use_attachment_pages', '1' );
+	}
+}
```
Comment (azaozz) [inline src/wp-admin/includes/upgrade.php]: Don't think it is a good idea to store an option in the DB. This should probably be a filter. Also seems that the front-end visibility of attachment post templates should be by template type, and best would be to be controlled by the theme.

File: src/wp-admin/options-media.php [hunk 2]
```diff
@@ -169,6 +169,18 @@
...
+<tr>
+<td class="td-full">
+<label for="wp_media_use_attachment_pages">
+<input name="wp_media_use_attachment_pages" type="checkbox" id="wp_media_use_attachment_pages" value="1"<?php checked( '1', get_option( 'wp_media_use_attachment_pages' ) ); ?> />
+	<?php _e( 'Use attachment pages' ); ?>
+</label>
+</td>
+</tr>
+</table>
+
```
Comment (azaozz) [inline src/wp-admin/options-media.php]: Don't think it is a good idea to "globalize" this as a setting, and to add a checkbox about this in WP.

File: src/wp-admin/options.php [hunk 3]
```diff
@@ -135,6 +135,7 @@
 		'image_default_size',
 		'image_default_align',
 		'image_default_link_type',
+		'wp_media_use_attachment_pages',
```
Comment (azaozz) [inline src/wp-admin/options.php]: Same as above.

File: src/wp-includes/canonical.php [hunk 4]
```diff
@@ -543,6 +543,10 @@ function redirect_canonical( $requested_url = null, $do_redirect = true ) {
 		}
 	}
 
+	if ( is_attachment() && '1' !== get_option( 'wp_media_use_attachment_pages' ) ) {
+		$redirect_url = wp_get_attachment_url( get_query_var( 'attachment_id' ) );
+	}
+
```
Comment (azaozz) [inline src/wp-includes/canonical.php]: The redirect may need to consider the (front-end) visibility of the attachment post and eventually return a 404 or an appropriate error page if the current user is not allowed to see it.

File: src/wp-includes/link-template.php [hunk 5]
```diff
@@ -517,6 +517,11 @@ function get_attachment_link( $post = null, $leavename = false ) {
 		$link = home_url( '/?attachment_id=' . $post->ID );
 	}
 
+	$redirect_to_attachment_url = '1' !== get_option( 'wp_media_use_attachment_pages' );
+	if ( $redirect_to_attachment_url ) {
+		$link = wp_get_attachment_url( $post->ID );
+	}
+
```
Comment (azaozz) [inline src/wp-includes/link-template.php]: Same as above. May need to consider the "visibility" of the attachment post.

**Review by azaozz — COMMENTED** [review]
Already commented on https://core.trac.wordpress.org/ticket/57913, adding some inline comments here to clarify.

**Comment by SergeyBiryukov** [comment]
Thanks for the PR! Merged in [r56657](https://core.trac.wordpress.org/changeset/56657).

**Pull request closed without merging**